alf/algorithms/ddpg_algorithm.py

```diff
@@ -359,6 +360,15 @@ def calc_loss(self, info: DdpgInfo):
             priority=priority,
             extra=DdpgLossInfo(critic=critic_loss, actor=actor_loss.extra))
 
+    def preprocess_experience(self, exp, rollout_info, batch_info):
```

Comment:
Is it possible to move this to a centralized place (e.g. Agent), so we do not need to add this for each algorithm?

Reply:
Good point.  I've thought about it, but don't have a good way at the moment.  1) We could put it in the OffPolicyAlgorithm parent class.  But subclasses may override it and accidentally skip this preprocessing.  2) Put in Agent.  But then DdpgAlgo standalone won't work.  3) AlgorithmInfo namedtuples are algorithm specific.  Unless we extract some common parts out and put inside say OffPolicyAlgo, there will always be some algorithm specific change involved to utilize discounted_return.

Maybe let's think more about it.

Reply:
A cleaner way is to extract the preprocess part into a Algorithm wrapper, containing the ddpg or sac algorithm as a sub-algorithm.  This will make it easy to configure any new algorithm.  Repurposing any existing config file with just command line args would be more difficult, and we'd probably have to create a separate config file for any runs using discounted_return.

@Haichao-Zhang If you are Ok with this, I can handle it perhaps in another PR.

Reply:
I think it's ok to put it in Agent. The intention of Agent is an integration holder for different algorithms.

Reply:
Sounds good.  Done (only when AlgoInfo has batch_info field).